sys/include/net/netbuf.h

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2019 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    net_netbuf Network buffer abstraction
+ * @ingroup     net
+ * @brief       Provides an abstraction for network stack buffers
+ * @{
+ *
+ * @file
+ * @brief       Network buffer abstraction definition
+ *
+ * @author  Jose I. Alamos <jose.alamos@haw-hamburg.de>
+ */
+#ifndef NET_NETBUF_H
+#define NET_NETBUF_H
+
+#include <stdbool.h>
+#include "iolist.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Get a network buffer with a given size.
+ *
+ * @note    Supposed to be implemented by the networking module.
+ * @note    @p ctx must be NOT NULL if the function returns a valid pointer.
```

Comment:
You can use `@post` instead of `@note` when declaring a postcondition.

Reply:
Isn't this more of a pre-condition: `ctx != NULL`? Making the validity of `ctx` dependent of the return value makes the function somewhat of a crystal ball otherwise ;-).

Reply:
What I was trying to say here is: there should always be allocation context if this function returns a pointer to a buffer.
If the `ctx` is set to NULL, later the `netif_recv` function has no way to know if the packet was allocated in the stack or a framebuffer

Reply:
it would be possible to avoid this is we set an extra variable in the `netif_pkt_t` that indicates the source of the pkt (framebuffer or stack). But I'm not sure if it's the best way to do it

Reply:
> What I was trying to say here is: there should always be allocation context if this function returns a pointer to a buffer.

Is there? What about stacks that don't need a context? Do they then need to set some bogus context just to make `netbuf` happy?

Reply:
IMHO, netif shouldn't touch the netbuf context in either case (just call `netbuf_free()` with it) this setting this post-condition is not needed.

Reply:
But then how do you tell `netif_recv` that the packet was not allocated by the stack, so there's no need to free? (and that's not required to remove the MAC header)

Reply:
then we would need something like https://github.com/RIOT-OS/RIOT/pull/12741#discussion_r349075967

Reply:
> But then how do you tell `netif_recv` that the packet was not allocated by the stack, so there's no need to free? (and that's not required to remove the MAC header)

Just call `netif_free` always. The stack knows how to handle the context (or it is just a NOP).

> (and that's not required to remove the MAC header)

If the MAC header is not parsed / marked, there is no need to remove it, right?

Reply:
> hen we would need something like [#12741 (comment)](https://github.com/RIOT-OS/RIOT/pull/12741#discussion_r349075967)

I don't see the relation between freeing memory regions and a source address....

Reply:
> Just call netif_free always. The stack knows how to handle the context (or it is just a NOP).

The problem is, is not ensured that `netbuf_get` is called before calling `netif_recv`.

E.g for the ESP WIFI:
```c
void esp_wifi_phy_indication(void *psdu, size_t psdu_len)
{
    /* No need to call netbuf_get since the psdu is already allocated somewhere... */
    ieee80211_phy_t phy;
    phy.psdu = psdu;
    phy.psdu_len = psdu_len;
    ...
    ieee80211_phy_ind(&phy);
}
```
and then in netif_recv:
```c
void netif_recv(...)
{
    if(packet_was_allocated_by_stack() ) {
        pkt = mark_and_remove_gnrc_pktsnip_mac_header();
    }
    else {
        /* the packet is in "some" buffer that has no relation with the stack */
        /* for GNRC we need to copy it to a gnrc_pktbuf anyway... */
        pkt = gnrc_pktsnip_add(netif_pkt->msdu.iol_base, netif_pkt->msdu.iol_len, GNRC_NETTYPE_UNDEF);
    }
        build_and_append_gnrc_netif_hdr(pkt)
        dispatch_packet(pkt);
}
```

How would you implement this `packet_was_allocated_by_stack()` function?.

Reply:
> I don't see the relation between freeing memory regions and a source address....

This is not source address, this is the source of the allocation (stack, framebuffer, etc)

Reply:
> How would you implement this packet_was_allocated_by_stack() function?.

There are two approches:

1. The stack just knows if a packet is from its own buffer.
    - **Example 1:** `gnrc_pktbuf_static` has a currently internal function `_pktbuf_contains()` to determine if a packet is within the packet buffer https://github.com/RIOT-OS/RIOT/blob/178d0aace20784e69a641c546c845fc8dce0fae1/sys/net/gnrc/pktbuf_static/gnrc_pktbuf_static.c#L58-L61
    - **Example 2:** The packet is stored in a global buffer, the boundaries of that buffer can be determined
2. The indication provides this information (e.g. in a flag) independent of `ctx`.

(1) has the disadvantage that is not easy to implement in some cases (e.g. `gnrc_pktbuf_malloc`, (2) has the disadvantage that a driver implementer can easily forget to set this flag (but the same is possible in the stack's implementation of `netbuf_alloc()` with the `ctx`).

The disadvantage of (1) is pretty huge, so we should disregard it. (2) could be mitigated by having an `ieee80211_phy_init()` function, e.g.:

```C
bool ieee80211_phy_init(ieee80211_phy_t *phy, void *psdu, size_t psdu_len)
{
     memset(phy, 0, sizeof(ieee80211_phy_t));
     if (DEVICE_PROVIDES_NETBUF && (psdu != NULL)) {
         phy->psdu = psdu;
         phy.psdu_len = psdu_len;
         phy->flags |= IEEE80211_PHY_BUF_PROVIDED_BY_DEV;
     }
     else {
         phy->psdu = netbuf_alloc(psdu_len, &phy->ctx);
     }
     return (phy->psdu != NULL);
}
```

In any case I would make such a functionality compile-time optimized (`packet_was_allocated_by_stack()` returns always true, if there is no device that behaves in that way). I drafted this above by using the macro `DEVICE_PROVIDES_NETBUF` in `ieee80211_phy_init_psdu()` as well.